zigbeetuya/integrationpluginzigbeetuya.cpp

```diff
@@ -309,6 +321,80 @@ void IntegrationPluginZigbeeTuya::setupThing(ThingSetupInfo *info)
+                DpValue dpValue = DpValue::fromData(frame.payload);
+
+                switch (dpValue.dp()) {
+                case LUMINANCE_MOTION_SENSOR_DP_PRESENCE:
+                    qCDebug(dcZigbeeTuya()) << "presence changed:" << dpValue;
+                    thing->setStateValue(motionSensorIsPresentStateTypeId, dpValue.value().toInt() == 0);
+                    break;
+                case LUMINANCE_MOTION_SENSOR_DP_BATTERY:
+                    qCDebug(dcZigbeeTuya()) << "Battery changed:" << dpValue;
+                    thing->setStateValue(motionSensorBatteryLevelStateTypeId, dpValue.value().toInt());
+                    thing->setStateValue(motionSensorBatteryCriticalStateTypeId, dpValue.value().toInt() < 5);
+                    break;
+                case LUMINANCE_MOTION_SENSOR_DP_SENSITIVITY:
+                    qCDebug(dcZigbeeTuya()) << "Sensitivity:" << dpValue << thing->setting(motionSensorSettingsSensitivityParamTypeId);
+                    thing->setSettingValue(motionSensorSettingsSensitivityParamTypeId, dpValue.value().toInt() + 1);
+                    break;
+                case LUMINANCE_MOTION_SENSOR_DP_KEEP_TIME: {
+                    qCDebug(dcZigbeeTuya()) << "timeout changed:" << dpValue << thing->setting(motionSensorSettingsTimeoutParamTypeId);
+                    QHash<int, int> map = {{0, 10}, {1, 30}, {2, 60}, {3, 120}};
+                    thing->setSettingValue(motionSensorSettingsTimeoutParamTypeId, map.value(dpValue.value().toInt()));
+                    break;
+                }
+                case LUMINANCE_MOTION_SENSOR_DP_ILLUMINANCE:
+                    qCDebug(dcZigbeeTuya()) << "illuminance changed:" << dpValue;
+                    thing->setStateValue(motionSensorLightIntensityStateTypeId, dpValue.value().toDouble());
+                    break;
+
+                default:
+                    qCWarning(dcZigbeeTuya()) << "Unhandled data point" << dpValue;
+                }
+
+            } else {
+                qCWarning(dcZigbeeTuya()) << "Unhandled presence sensor cluster command:" << frame.header.command;
+            }
+        });
+
+        connect(thing, &Thing::settingChanged, cluster, [cluster, thing, this](const ParamTypeId &settingTypeId, const QVariant &value) {
+            DpValue dp;
+
+            if (settingTypeId == motionSensorSettingsTimeoutParamTypeId) {
+                QHash<int, int> map = {{0, 10}, {1, 30}, {2, 60}, {3, 120}};
+                dp = DpValue(LUMINANCE_MOTION_SENSOR_DP_KEEP_TIME, DpValue::TypeEnum, map.key(value.toUInt()), 4, m_seq++);
+            }
+            if (settingTypeId == motionSensorSettingsSensitivityParamTypeId) {
+                dp = DpValue(LUMINANCE_MOTION_SENSOR_DP_SENSITIVITY, DpValue::TypeEnum, value.toUInt(), 4, m_seq++);
+            }
+            qCDebug(dcZigbeeTuya()) << "setting" << thing->thingClass().settingsTypes().findById(settingTypeId).name() << dp << dp.toData().toHex();
```

Comment:
Ever since this is a sleepy device I guess, wouldn't it be good to queue the settings requests until you revceive data from the device? I'm still thinking about doing such things in the network class, so we don't have to care about it here...but changing the settings could end up into just a timeout and it remains undone?

Reply:
yes, fair point... but indeed that should probably be done in the library instead of doing that all over the place.